drivers/hmis/app/graphql/types/hmis_schema/has_ce_opportunities.rb

```diff
@@ -0,0 +1,52 @@
+###
+# Copyright 2016 - 2025 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# frozen_string_literal: true
+
+module Types
+  module HmisSchema
+    module HasCeOpportunities
+      extend ActiveSupport::Concern
+
+      include ::Hmis::Concerns::HmisArelHelper
+
+      class_methods do
+        def ce_opportunities_field(name = :ce_opportunities, description = nil, filter_args: {}, **override_options, &block)
+          default_field_options = { type: Types::HmisSchema::CeOpportunity.page_type, null: false, description: description }
+          field_options = default_field_options.merge(override_options)
+          field(name, **field_options) do
+            argument :sort_order, Types::HmisSchema::CeOpportunitySortOption, required: false
+            filters_argument HmisSchema::CeOpportunity, **filter_args
+            instance_eval(&block) if block_given?
+          end
+        end
+      end
+
+      included do
+        def resolve_ce_opportunities(scope = object.ce_opportunities, **args)
+          scoped_ce_opportunities(scope, **args)
+        end
+      end
+
+      private
+
+      def scoped_ce_opportunities(scope, user: current_user, sort_order: :date_available_earliest_first, filters: nil)
+        raise unless Hmis::Ce.configuration.enabled? # TODO(#7506) permissions
+
+        scope = scope.viewable_by(user)
+
+        scope = scope.where(status: filters&.status) if filters.respond_to?(:status) && filters&.status.present?
```

Comment:
To discuss: I don't really love this messy `filter.respond_to?` check peppered in everywhere. I was wondering if we have a common pattern for dealing with this; maybe that's what the `yield_self` in `EnrollmentFilter` helps with? https://github.com/greenriver/hmis-warehouse/blob/stable/drivers/hmis/app/models/hmis/filter/enrollment_filter.rb

Reply:
I think that's what the `with_filter` helper does in the filter classes (like the one you linked). It might be good to follow the pattern of having a dedicated class for filtering, since this has a lot of filter capabilities. It lets us test filters in isolation without graphql.

Reply:
Thanks! Hm do you have any examples in the code you can link me to where we are testing filters in isolation without graphql? I wasn't finding any instances of `apply_filters` in the spec files.

Reply:
Ha I guess just because we _can_ in theory  doesn't mean we are 👀